### PR TITLE
Simplify and clarify FPU decimation loop

### PR DESCRIPTION
**Pull request by kcgen**
Replaces the four integer operations with two calls to std::div, which lets the compiler optimize the resulting assembly calls from:

## Original integer calls

**uint64_t temp = rndint / 10;**
``` asm
        mov     rax, QWORD PTR [rbp-8]
        movabs  rdx, -3689348814741910323
        mul     rdx
        mov     rax, rdx
        shr     rax, 3
        mov     QWORD PTR [rbp-16], rax
```

**uint8_t p = static_cast<uint8_t>(rndint % 10);**
``` asm
        mov     rcx, QWORD PTR [rbp-8]
        movabs  rdx, -3689348814741910323
        mov     rax, rcx
        mul     rdx
        shr     rdx, 3
        mov     rax, rdx
        sal     rax, 2
        add     rax, rdx
        add     rax, rax
        sub     rcx, rax
        mov     rdx, rcx
        mov     BYTE PTR [rbp-17], dl
```

**rndint = temp / 10;**
``` asm
        mov     rax, QWORD PTR [rbp-16]
        movabs  rdx, -3689348814741910323
        mul     rdx
        mov     rax, rdx
        shr     rax, 3
        mov     QWORD PTR [rbp-8], rax
```

**p |= (static_cast<uint8_t>(temp % 10)) << 4;**
``` asm
        mov     rcx, QWORD PTR [rbp-16]
        movabs  rdx, -3689348814741910323
        mov     rax, rcx
        mul     rdx
        shr     rdx, 3
        mov     rax, rdx
        sal     rax, 2
        add     rax, rdx
        add     rax, rax
        sub     rcx, rax
        mov     rdx, rcx
        mov     eax, edx
        movzx   eax, al
        sal     eax, 4
        mov     edx, eax
        movzx   eax, BYTE PTR [rbp-17]
        or      eax, edx
        mov     BYTE PTR [rbp-17], al
```

## standard library calls to DIV

**const lldiv_t div10 = lldiv(rndint, 10);**
``` asm
        mov     rax, QWORD PTR [rbp-8]
        mov     esi, 10
        mov     rdi, rax
        call    lldiv
        mov     QWORD PTR [rbp-48], rax
        mov     QWORD PTR [rbp-40], rdx
```

**const lldiv_t div100 = lldiv(div10.quot, 10);**
``` asm
        mov     rax, QWORD PTR [rbp-48]
        mov     esi, 10
        mov     rdi, rax
        call    lldiv
        mov     QWORD PTR [rbp-64], rax
        mov     QWORD PTR [rbp-56], rdx
```

**p = static_cast<uint8_t>(div10.rem) | (static_cast<uint8_t>(div100.rem) << 4);**

``` asm
        mov     rax, QWORD PTR [rbp-40]
        mov     edx, eax
        mov     rax, QWORD PTR [rbp-56]
        movzx   eax, al
        sal     eax, 4
        or      eax, edx
        mov     BYTE PTR [rbp-17], al
```

**rndint = div100.quot;**
``` asm
        mov     rax, QWORD PTR [rbp-64]
        mov     QWORD PTR [rbp-8], rax
```

## References
  - Assembly comparison: https://godbolt.org/z/Tx5zYc
  - Runtime comparison: https://onlinegdb.com/HyZTi5vSD

### CODE REVIEW

**Comment by kcgen** [comment]
Runnning OLE2DSP (from [here](https://github.com/dosbox-staging/dosbox-staging/issues/610)) produces the same result as well.

**Comment by kcgen** [comment]
Given `lldiv` is only offered in C++11, this is likely why upstream was limited to using separate division and remainder calls.

**Comment by kcgen** [comment]
In GodBolt, if you switch the output to the RISC platform, you can see the four separate math calls:

``` asm
        divu    a5,a4,a5
        ...
        remu    a5,a4,a5
        ...
        divu    a5,a4,a5
        ...
        remu    a5,a4,a5
```

versus two hardware-calls to lldiv:

``` asm
        call    lldiv
        call    lldiv
```

**Comment by kcgen** [comment]
Testing a debug build with host CPU locked 800 Mhz to magnify the differences, running dosbox-staging at a fixed cycles of 3000 and running the Landmark CPU and FPU benchmark, the original code uses slightly more of the host's CPU.

### Before

Ranges between mid 77% and 78%



### After

Ranges between mid 76% and low 77%

**Comment by grapeli** [comment]
The result for OLE2DSP is correct after this [commit](https://github.com/dosbox-staging/dosbox-staging/commit/aae120b1bc7b6728597f66b0aabbab0c1841e40b).

Change config to:
```
[cpu]
core = normal
```
and now validate the QUADTEST.

**Comment by kcgen** [comment]
Thanks @grapeli!
Would you be able to test this PR and update your comment? (I wasn't sure from your comment exactly what pertained to the original commit vs this PR's refinement). Much thanks!

**Comment by kcgen** [comment]
Thanks for clarifying and confirming @grapeli; and apologies for not following your original wording :-)

File: src/fpu/fpu_instructions.h [hunk 1]
```diff
@@ -299,11 +299,12 @@ static void FPU_FBST(PhysPt addr) {
 
 	//numbers from back to front
 	for(Bitu i=0;i<9;i++){
-		Bit64u temp = rndint / 10;
-		Bit8u p = static_cast<Bit8u>(rndint % 10);
-		rndint = temp / 10;
-		p |= (static_cast<Bit8u>(temp % 10)) << 4;
-		mem_writeb(addr++,p);
+		const lldiv_t div10 = lldiv(rndint, 10);
```
Comment (dreamer) [inline src/fpu/fpu_instructions.h:302]: I think we need to assure, that `unsigned long long` is indeed the same type as `uint64_t`. Static assert before line 301? (You could replace `Bitu i` with simple `int i` at the same time, although it does not really matter - compiler will probably do it anyway while optimizing).

When playing with this code in Compiler Explorer I noticed, that if `mem_writeb` is moved out of the loop, the compiler nicely unrolls the loop (when compiled with -O3). Maybe it would result in faster code or maybe not (I can't say without doing microbenchmarks) - leaving it as an idea for future improvement, but I think it's out of the scope of this PR :).

**Review by dreamer — CHANGES_REQUESTED** [review]
Clever change :) I looked up StackOverflow and consensus seems to be that in most situations modern C++ compilers are good enough and will optimize code better without programmer worrying about `div`/`ldiv`/`lldiv` - but in this case it seems to be beneficial.

Small gripe about the commit message: it's a little bit unfair to compare both versions using -O0 - I suggest rewriting them after compiling with -O3 - old asembler version seems to be much smaller (but version with `lldiv` still generates smaller assembly; but I don't know if faster).

Tests for CPU usage were done with -O3 or with debug builds?

**Comment by kcgen** [comment]
Thanks for the quick review @dreamer.

Will add the static asset to check the type equivalence; nice suggestion.

Yes, the benchmark was deliberately done with a debug build and at 800 MHz because both were needed to magnify the host cpu usage difference to the point where I could measure it.

At higher optimization levels and/or faster processor speeds, the cpu usage bled into eachother far enough that they became too similar to differentiate. 

But yes, the goal of the change was more about clarifying  using better names to show that we're dividing by 10 and then 100, as well as use div(), as that's a common call for when you need both the quotient and remainder for the same division.

(Likewise, it takes any guesswork out of hoping the compiler can connect the dots an optimize the separate calls into one)

**Comment by MasterO2** [comment]
For those of us who aren't programmers, what does this do for dosbox-staging exactly?

**Comment by kcgen** [comment]
@MasterO2, from a user perspective:

1. Functionally wise, the FPU will produce identical results as the prior code. 

2. Performance wise, this change guarantees that half the number of host-side integer operations are required to perform this FPU operation in dosbox compared to the prior code.

"Guarantee", in this case, meaning that there is some chance that the previous code /might/ be optimized by the compiler to similarly cut the operations in half - but in some cases, it might not.  For desktop PCs, this is of no concern. But for more obscure slower hardware (ie: Pi 1, my 32bit PPC 7400, and so on) the benefit will be measurable and likely let you run games with slightly more cycles than without.

File: src/fpu/fpu_instructions.h (hunk 1 above)
Comment (kcgen) [reply src/fpu/fpu_instructions.h:304]: Added the static assert.

Comment (kcgen) [reply src/fpu/fpu_instructions.h:304]: > I noticed, that if mem_writeb is moved out of the loop, the compiler nicely unrolls the loop (when compiled with -O3). Maybe it would result in faster code or maybe not (I can't say without doing microbenchmarks) 

This means it's storing all 9 intermediate results from each loop pass; I guess we could store them in a temporary array, then write at the end (would be a tad uglier; but probably faster because it could SIMD the for loop with 18 calls to lldiv and stores).

**Comment by dreamer** [comment]
@kcgen 
> I don't recall if godbolt uses some level of optimization, but many times the compiler wasn't getting it on the various platforms.

godbolt uses the flags you tell it to use ;) compiler options can be bassed in a text field above the assembler pane.

I think one additional note worth noting: this change affects only builds that opt-out or can't use FPU emulators written in x86 or x86-64 assemblers, therefore this change affects e.g.:

- Linux ARM builds
- Windows x64 builds
- (future) macOS ARM builds

but does not affect:

- Linux x86_64 builds
- Windows x32 builds
- (current) macOS x86-64 builds

I don't feel like doing in-depth performance testing and comparing on all platforms, so just as a proxy I looked what assembler will be generated on ARM (because every performance gain for our RPi users is important):

### Before:
*using printf instead of mem_writeb to assure loop won't be unrolled*
```
.L2:
        mov     r2, #10
        mov     r3, #0
        mov     r0, r8
        mov     r1, r7
        bl      __aeabi_uldivmod
        mov     r2, #10
        mov     r6, r0
        mov     r5, r1
        mov     r0, r8
        mov     r1, r7
        mov     r3, #0
        bl      __aeabi_uldivmod
        mov     r3, #0
        mov     r4, r2
        mov     r0, r6
        mov     r2, #10
        mov     r1, r5
        bl      __aeabi_uldivmod
        mov     r2, #10
        mov     r8, r0
        mov     r7, r1
        mov     r0, r6
        mov     r1, r5
        mov     r3, #0
        bl      __aeabi_uldivmod
        orr     r4, r4, r2, lsl #4
        mov     r0, r10
        and     r1, r4, #255
        bl      printf
        subs    r9, r9, #1
        bne     .L2
```

### After
```
.L2:
        mov     r2, r8
        mov     r3, r5
        add     r0, sp, #8
        stm     sp, {r6-r7}
        bl      lldiv
        add     r3, sp, #8
        ldmia   r3, {r2-r3}
        add     r0, sp, #24
        stm     sp, {r6-r7}
        bl      lldiv
        ldrb    r3, [sp, #32]   @ zero_extendqisi2
        ldrb    r1, [sp, #16]   @ zero_extendqisi2
        orr     r1, r1, r3, lsl #4
        mov     r0, r9
        and     r1, r1, #255
        ldr     r8, [sp, #24]
        ldr     r5, [sp, #28]
        bl      printf
        subs    r4, r4, #1
        bne     .L2
```

So: on ARM we had 4 `bl` instructions related to floating-point calculations, and now we'll have 2. 
When comparing to x86-64 we had 0 `call` instructions and longer assembly; now we have 1 `call` and shorter assembly.

### Summary

In plain words: on x86-64 we can't be sure this change improves things (it will depend on specific CPU), but it affects only Windows x64 builds. It might be faster, might be slower - either way, the difference will be hard to measure on modern x86-64 CPUs.

On the other hand, on ARM CPUs (where we need every boost we can get for our RPi users), this is clearly beneficial change.

**Review by dreamer — APPROVED** [review]
I will slightly adjust the static_assert after merging the code, no point in going through another review round for this.

Thanks! :)

**Comment by dreamer** [comment]
Merged in: f368bbf665823d1fe74ec9ef03f03572975f0441
My adjustment: 92969079149e5d2aba12563254d4e3e9fb7aea8e

@kcgen Thanks! :)

**Comment by kcgen** [comment]
Thank you @dreamer for the excellent summary of this change's impact/benefits, as well as the improved `static_assert` that directly tests the variable involved.
